internal/testutils/stats/test_metrics_recorder.go

```diff
@@ -41,17 +43,74 @@ type TestMetricsRecorder struct {
 	intHistoCh   *testutils.Channel
 	floatHistoCh *testutils.Channel
 	intGaugeCh   *testutils.Channel
+
+	// The most recent update for each metric name.
+	mu   sync.Mutex
+	data map[estats.Metric]float64
 }
 
 func NewTestMetricsRecorder(t *testing.T, metrics []string) *TestMetricsRecorder {
-	return &TestMetricsRecorder{
+	tmr := &TestMetricsRecorder{
 		t: t,
 
-		intCountCh:   testutils.NewChannelWithSize(10),
-		floatCountCh: testutils.NewChannelWithSize(10),
-		intHistoCh:   testutils.NewChannelWithSize(10),
-		floatHistoCh: testutils.NewChannelWithSize(10),
-		intGaugeCh:   testutils.NewChannelWithSize(10),
+		intCountCh:   testutils.NewChannelWithSize(1000),
+		floatCountCh: testutils.NewChannelWithSize(1000),
+		intHistoCh:   testutils.NewChannelWithSize(1000),
+		floatHistoCh: testutils.NewChannelWithSize(1000),
+		intGaugeCh:   testutils.NewChannelWithSize(1000),
+
+		data: make(map[estats.Metric]float64),
+	}
+
+	for _, metric := range metrics {
+		tmr.data[estats.Metric(metric)] = 0
+	}
+
+	return tmr
+}
+
+// AssertDataForMetric asserts data is present for metric. The zero value in the
+// check is equivalent to unset.
+func (r *TestMetricsRecorder) AssertDataForMetric(metricName string, wantVal float64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.data[estats.Metric(metricName)] != wantVal {
+		r.t.Fatalf("Unexpected data for metric %v, got: %v, want: %v", metricName, r.data[estats.Metric(metricName)], wantVal)
+	}
+}
+
+// AssertEitherDataForMetric asserts either data point is present for metric.
+// The zero value in the check is equivalent to unset.
+
+func (r *TestMetricsRecorder) AssertEitherDataForMetric(metricName string, wantVal1 float64, wantVal2 float64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.data[estats.Metric(metricName)] != wantVal1 && r.data[estats.Metric(metricName)] != wantVal2 {
+		r.t.Fatalf("Unexpected data for metric %v, got: %v, want: %v or %v", metricName, r.data[estats.Metric(metricName)], wantVal1, wantVal2)
+	}
+}
+
+// PollForDataForMetric polls the metric data for the want. Fails if context
+// provided expires before data for metric is found.
+func (r *TestMetricsRecorder) PollForDataForMetric(ctx context.Context, metricName string, wantVal float64) {
+	for ; ctx.Err() == nil; <-time.After(time.Millisecond) {
+		r.mu.Lock()
+		if r.data[estats.Metric(metricName)] == wantVal {
+			r.mu.Unlock()
+			break
```

Comment:
Won't this fail if you ever use it because of the `Fatalf`?  You probably want `return`, but if nothing is using it, then delete it?

Reply:
Oh whoops, switched to return.